### PR TITLE
Fixed #27225 -- Added "Age" header when fetching cached responses.

### PR DESCRIPTION
**Pull request by e11bits**
This addresses Ticket 27225. I took the PR #7246 from @renskiy and rebased it to the current main branch.

Tests pass locally and I added just a minor change to `hello_world_view` .  I haven't yet added any missing changes recommended in #7246 by @timgraham , so I create this as a draft PR first.

This is my fourth PR for django, so please advise as you see fit.

### CODE REVIEW

**Comment by e11bits** [comment]
In addition to the original PR I did the following:

- squashed to commits
- Removed the change in `django/utils/cache.py` . I also couldn't see why this change would be necessary. see original comment https://github.com/django/django/pull/7246#discussion_r81975841
- Added a comment and renamed some of the variables in `django/middleware/cache.py` to make it more obvious how the calculation is done. see original comment https://github.com/django/django/pull/7246#discussion_r81975935
- Changed the comment in the test case to explain the value of the `Age: 1` header entry. see original comment https://github.com/django/django/pull/7246#discussion_r81976035

**Comment by e11bits** [comment]
Ok, now I know, why the former code in `django/utils/cache.py` was introduced by the original PR. This is due to another bug for which probably have to open a new ticket first.

**Comment by e11bits** [comment]
This is currently blocked by #17776

**Comment by MiirzaBaig** [comment]
There is an error in cache middleware system.

**Comment by e11bits** [comment]
Not blocked by https://github.com/django/django/pull/17776 anymore

File: tests/cache/tests.py [hunk 1]
```diff
@@ -2477,8 +2477,8 @@ class PrefixedCacheI18nTest(CacheI18nTest):
     pass
 
 
-def hello_world_view(request, value):
-    return HttpResponse("Hello World %s" % value)
+def hello_world_view(request, value=None):
+    return HttpResponse("Hello World %s" % (value or ""))
```
Comment (felixxm) [inline tests/cache/tests.py:2481]: I'd revert this change and pass `value` to the `view()` calls:
```diff
diff --git a/tests/cache/tests.py b/tests/cache/tests.py
index ce02872d4e..40d07fd45e 100644
--- a/tests/cache/tests.py
+++ b/tests/cache/tests.py
@@ -2477,8 +2477,8 @@ class PrefixedCacheI18nTest(CacheI18nTest):
     pass
 
 
-def hello_world_view(request, value=None):
-    return HttpResponse("Hello World %s" % (value or ""))
+def hello_world_view(request, value):
+    return HttpResponse("Hello World %s" % value)
 
 
 def csrf_view(request):
@@ -2760,7 +2760,7 @@ class CacheMiddlewareTest(SimpleTestCase):
         with mock.patch.object(
             time, "time", return_value=1468749600
         ):  # Sun, 17 Jul 2016 10:00:00 GMT
-            response = view(request)
+            response = view(request, 1)
             response.close()
             self.assertIn("Expires", response)
             self.assertEqual(response["Expires"], "Sun, 17 Jul 2016 10:00:02 GMT")
@@ -2773,7 +2773,7 @@ class CacheMiddlewareTest(SimpleTestCase):
         with mock.patch.object(
             time, "time", return_value=1468749601
         ):  # Sun, 17 Jul 2016 10:00:01 GMT
-            response = view(request)
+            response = view(request, 1)
             response.close()
             self.assertIn("Expires", response)
             self.assertEqual(response["Expires"], "Sun, 17 Jul 2016 10:00:02 GMT")

```

**Review by felixxm — COMMENTED** [review]
@e11bits Thanks :+1: Can you correct the commit message? This PR is not about `max-age`.

File: django/middleware/cache.py [hunk 2]
```diff
@@ -171,6 +174,17 @@ def process_request(self, request):
             request._cache_update_cache = True
             return None  # No cache information available, need to rebuild.
 
+        # Derive the age estimation of the cached entry using max_age and the Expires
+        # timestamp. max_age_seconds is None if CACHE_MIDDLEWARE_SECONDS is
+        # misconfigured to be a float.
+        max_age_seconds = get_max_age(response)
+        if max_age_seconds is not None:
+            expires_timestamp = parse_http_date(response["Expires"])
```
Comment (kezabelle) [inline django/middleware/cache.py:182]: 🤔 I'm wondering out loud, mostly. Is this a scenario where we care/need to handle the possibility of an exception being raised? There's other places which act similarly upon "response" headers[^1] and do so by leveraging `parse_http_date_safe`[^2] to avoid crashing, but it's decidedly unclear to me as to whether that's relevant here...

Just flagging it for consideration, as such.

[^1]: "last modified" & "if modified since"
[^2]: which just catches and returns `None`

Comment (e11bits) [reply django/middleware/cache.py:182]: I thought about that too, but came to the conclusion, that extra caution is not required here, since the `Expires` response header is set by a django cache. 

`response.headers["Expires"] = http_date(time.time() + cache_timeout)`

Two cases you mentioned in `cache.py` are dealing with request headers `If-Modified-Since` and `If-Unodified-Since` that could likely contain badly formatted date strings and need `parse_http_date_safe`

`Last-Modified` from a request will be saved in the cache as is and could be badly formatted as well and is used for a conditional response.

Comment (felixxm) [reply django/middleware/cache.py:182]: > ...since the `Expires` response header is set by a django cache.

Yes, but only if it's not already set.

Comment (e11bits) [reply django/middleware/cache.py:182]: > > ...since the `Expires` response header is set by a django cache.
> 
> Yes, but only if it's not already set.

True. I missed `Expires` being part of `_not_modified`.

Comment (e11bits) [reply django/middleware/cache.py]: > What do you think about the following change?

Looks good. Made the change.

**Comment by felixxm** [comment]
@e11bits Thanks :+1:
